code/datums/actions/mobs/charge.dm

```diff
@@ -0,0 +1,228 @@
+		return
+	if(!isliving(A))
+		return
+	var/mob/living/L = A
+	L.visible_message("<span class='danger'>[source] slams into [L]!</span>", "<span class='userdanger'>[source] tramples you into the ground!</span>")
+	source.forceMove(get_turf(L))
+	L.apply_damage(damage_dealt, BRUTE, wound_bonus = CANT_WOUND)
+	playsound(get_turf(L), 'sound/effects/meteorimpact.ogg', 100, TRUE)
+	shake_camera(L, 4, 3)
+	shake_camera(source, 2, 3)
+
+/datum/action/cooldown/mob_cooldown/charge/triple_charge
+	name = "Triple Charge"
+	desc = "Allows you to charge three times at a chosen position."
+	charge_delay = 0.6 SECONDS
+
+/datum/action/cooldown/mob_cooldown/charge/triple_charge/charge_sequence(atom/movable/charger, atom/target_atom, delay, past)
+	for(var/i in 0 to 2)
+		do_charge(owner, target_atom, charge_delay - 2 * i, charge_past)
+
+/datum/action/cooldown/mob_cooldown/charge/hallucination_charge
+	name = "Hallucination Charge"
+	icon_icon = 'icons/effects/bubblegum.dmi'
+	button_icon_state = "smack ya one"
+	desc = "Allows you to create hallucinations that charge around your target."
+	cooldown_time = 2 SECONDS
+	charge_delay = 0.6 SECONDS
+	/// The damage the hallucinations in our charge do
+	var/hallucination_damage = 15
+	/// Check to see if we are enraged, enraged ability does more
+	var/enraged = FALSE
+	/// Check to see if we should spawn blood
+	var/spawn_blood = FALSE
+
+/datum/action/cooldown/mob_cooldown/charge/hallucination_charge/charge_sequence(atom/movable/charger, atom/target_atom, delay, past)
+	if(!enraged)
+		hallucination_charge(target_atom, 6, 8, 0, 6, TRUE)
+		StartCooldown(cooldown_time * 0.5)
+		return
+	for(var/i in 0 to 2)
+		hallucination_charge(target_atom, 4, 9 - 2 * i, 0, 4, TRUE)
+	for(var/i in 0 to 2)
+		do_charge(owner, target_atom, charge_delay - 2 * i, charge_past)
+
+/datum/action/cooldown/mob_cooldown/charge/hallucination_charge/do_charge(atom/movable/charger, atom/target_atom, delay, past)
+	. = ..()
+	if(charger != owner)
+		qdel(charger)
+
+/datum/action/cooldown/mob_cooldown/charge/hallucination_charge/proc/hallucination_charge(atom/target_atom, clone_amount, delay, past, radius, use_self)
+	var/starting_angle = rand(1, 360)
+	if(!radius)
+		return
+	var/angle_difference = 360 / clone_amount
+	var/self_placed = FALSE
+	for(var/i = 1 to clone_amount)
+		var/angle = (starting_angle + angle_difference * i)
+		var/turf/place = locate(target_atom.x + cos(angle) * radius, target_atom.y + sin(angle) * radius, target_atom.z)
+		if(!place)
+			continue
+		if(use_self && !self_placed)
+			owner.forceMove(place)
+			self_placed = TRUE
+			continue
+		var/mob/living/simple_animal/hostile/megafauna/bubblegum/hallucination/B = new /mob/living/simple_animal/hostile/megafauna/bubblegum/hallucination(place)
+		B.appearance = owner.appearance
+		B.name = "[owner]'s hallucination"
+		B.alpha = 127.5
+		B.move_through_mob = owner
+		B.spawn_blood = spawn_blood
+		INVOKE_ASYNC(src, .proc/do_charge, B, target_atom, delay, past)
```

Comment:
needs more descriptive of a variable name, avoid one letter variable names like these